lib/lambda_ethereum_consensus/ssz_ex/decode.ex

```diff
@@ -0,0 +1,379 @@
+    <<first_offset::integer-32-little, rest_bytes::bitstring>> = binary
+    num_elements = div(first_offset, @bytes_per_length_offset)
+
+    if Integer.mod(first_offset, @bytes_per_length_offset) != 0 ||
+         first_offset < @bytes_per_length_offset do
+      {:error, "InvalidListFixedBytesLen"}
+    else
+      with :ok <-
+             sanitize_offset(first_offset, nil, byte_size(binary), first_offset) do
+        decode_variable_list_elements(
+          num_elements,
+          rest_bytes,
+          inner_type,
+          first_offset,
+          binary,
+          first_offset,
+          []
+        )
+        |> Enum.reverse()
+        |> Utils.flatten_results()
+      end
+    end
+  end
+
+  defp decode_variable_list_elements(
+         1 = _num_elements,
+         _acc_rest_bytes,
+         inner_type,
+         offset,
+         binary,
+         _first_offset,
+         results
+       ) do
+    part = :binary.part(binary, offset, byte_size(binary) - offset)
+    [decode(part, inner_type) | results]
+  end
+
+  defp decode_variable_list_elements(
+         num_elements,
+         acc_rest_bytes,
+         inner_type,
+         offset,
+         binary,
+         first_offset,
+         results
+       ) do
+    <<next_offset::integer-32-little, rest_bytes::bitstring>> = acc_rest_bytes
+
+    with :ok <-
+           sanitize_offset(next_offset, offset, byte_size(binary), first_offset) do
+      part = :binary.part(binary, offset, next_offset - offset)
+
+      decode_variable_list_elements(
+        num_elements - 1,
+        rest_bytes,
+        inner_type,
+        next_offset,
+        binary,
+        first_offset,
+        [decode(part, inner_type) | results]
+      )
+    end
+  end
+
+  defp decode_variable_container(binary, module) do
+    schemas = module.schema()
+    fixed_length = get_fixed_length(schemas)
+
+    with :ok <- sanitize_offset(fixed_length, nil, byte_size(binary), nil),
+         <<fixed_binary::binary-size(fixed_length), variable_binary::bitstring>> = binary,
+         {:ok, fixed_parts, offsets, items_index} <-
+           decode_fixed_section(fixed_binary, schemas, fixed_length),
+         :ok <- check_first_offset(offsets, items_index, byte_size(binary)),
+         {:ok, variable_parts} <- decode_variable_section(binary, variable_binary, offsets) do
+      {:ok, struct!(module, fixed_parts ++ variable_parts)}
+    end
+  end
+
+  defp decode_fixed_container(binary, module) do
+    schemas = module.schema()
+    fixed_length = get_fixed_length(schemas)
+
+    with {:ok, fixed_parts, _offsets, items_index} <-
+           decode_fixed_section(binary, schemas, fixed_length),
+         :ok <- check_byte_len(items_index, byte_size(binary)) do
+      {:ok, struct!(module, fixed_parts)}
+    end
+  end
+
+  defp check_first_offset([{offset, _} | _rest], items_index, _binary_size) do
+    cond do
+      offset < items_index -> {:error, "OffsetIntoFixedPortion (#{offset})"}
```

Comment:
this seems inconsistent with other error messages. I would change the error message to something more readable